engine/graphics/src/vulkan/graphics_vulkan_device.cpp

```diff
@@ -74,36 +75,97 @@ namespace dmGraphics
         return num_attributes;
     }
 
-    VkResult DescriptorAllocator::Allocate(VkDevice vk_device, VkDescriptorSetLayout* vk_descriptor_set_layout, uint8_t setCount, VkDescriptorSet** vk_descriptor_set_out)
+    static VkResult AllocateDescriptorPool(DescriptorAllocator* allocator, VkDevice vk_device)
     {
-        assert(m_DescriptorMax >= (m_DescriptorIndex + setCount));
-        *vk_descriptor_set_out  = &m_Handle.m_DescriptorSets[m_DescriptorIndex];
-        m_DescriptorIndex      += setCount;
+        VkDescriptorPool pool_handle = VK_NULL_HANDLE;
+        VkResult res = CreateDescriptorPool(vk_device, allocator->m_DescriptorsPerPool, &pool_handle);
+
+        DescriptorPool new_pool    = {};
+        new_pool.m_DescriptorPool  = pool_handle;
+        new_pool.m_DescriptorCount = 0;
+
+        allocator->m_DescriptorPools.OffsetCapacity(1);
+        allocator->m_DescriptorPools.Push(new_pool);
+
+        return res;
+    }
+
+    static void AllocateDescriptorSets(DescriptorAllocator* allocator)
+    {
+        if (allocator->m_DescriptorSets == 0)
+        {
+            allocator->m_DescriptorSets = (VkDescriptorSet*) malloc(sizeof(VkDescriptorSet) * allocator->m_DescriptorSetMax);
+        }
+        else
+        {
+            allocator->m_DescriptorSets = (VkDescriptorSet*) realloc(allocator->m_DescriptorSets, sizeof(VkDescriptorSet) * allocator->m_DescriptorSetMax);
+        }
+    }
+
+    static VkResult Prepare(DescriptorAllocator* allocator, VkDevice vk_device, uint32_t num_descriptors, uint32_t num_sets)
+    {
+        VkResult res = VK_SUCCESS;
+
+        DescriptorPool& pool = allocator->m_DescriptorPools[allocator->m_DescriptorPoolIndex];
+
+        if ((allocator->m_DescriptorsPerPool - pool.m_DescriptorCount) < num_descriptors)
+        {
+            allocator->m_DescriptorPoolIndex++;
+            if (allocator->m_DescriptorPoolIndex >= allocator->m_DescriptorPools.Size())
+            {
+                res = AllocateDescriptorPool(allocator, vk_device);
+            }
+        }
```

Comment:
We now keep adding pools if we need to (as suggested by samsungs documentation). Previously we resized the pool by removing the old pool and replacing it with a new pool, but according to samsung this is the preferred way.